endatix/endatix
# feat: Form Submission Webhook

# Adding support for firing webhook on Form Submission

## Description
- Added firing web hooks on Form Submission event handler
- Added support for multiple Urls destination
- All web hooks are fired in background threads not blocking the UI of the operation
- Added resilience strategy via HTTP Resiliance libraries - https://learn.microsoft.com/en-us/dotnet/core/resilience/http-resilience?tabs=dotnet-cli. 
- Added strategy to send headers with information on EventName, Entity & Action
- Added configs to modify RetryAttempts, Timeouts, Concurrency and other options
- Added tests for the Endatix.Core part of the feature (Endatix.

>[!IMPORTANT]
> - Documentation  is out of scope as it's blocked by merging auth PR. Task has been created - https://github.com/endatix/endatix/issues/170
> - Unit Tests in for the WebHooks parts of the code located in Endatix.Infrastructure  is out of scope as it's blocked by merging auth PR. Task has been created - https://github.com/endatix/endatix/issues/171

### How To Test
1. Update `appSettings.Development.config` and make sure to add the "https://webhook.site/91b2292d-a73f-4dd3-8d48-5fe164c4d210" Url to the config part
```
 "WebHooks" : {
      "Events": {
        "FormSubmitted": {
          "IsEnabled": true,
          "WebHookUrls": ["https://webhook.site/91b2292d-a73f-4dd3-8d48-5fe164c4d210"]
        }
      }
    }
```
2. Run the project
3. Open the https://webhook.site/#!/view/91b2292d-a73f-4dd3-8d48-5fe164c4d210 Url to check for WebHooks
4. Send a submission to a form

### Web Hooks signature and behavior was modelled by following best practices showcased and implemented in the GitHub's WebHooks. More info here - https://docs.github.com/en/webhooks/webhook-events-and-payloads. Below is comparison of the payloads for reference:

### Sample Payload GitHub 

```
> X-GitHub-Delivery: 72d3162e-cc78-11e3-81ab-4c9367dc0958
> X-Hub-Signature: sha1=7d38cdd689735b008b3c702edd92eea23791c5f6
> X-Hub-Signature-256: sha256=d57c68ca6f92289e6987922ff26938930f6e66a2d161ef06abdf1859230aa23c
> User-Agent: GitHub-Hookshot/044aadd
> Content-Type: application/json
> Content-Length: 6615
> X-GitHub-Event: issues
> X-GitHub-Hook-ID: 292430182
> X-GitHub-Hook-Installation-Target-ID: 79929171
> X-GitHub-Hook-Installation-Target-Type: repository

> {
>   "action": "opened",
>   "issue": {
>     "url": "https://api.github.com/repos/octocat/Hello-World/issues/1347",
>     "number": 1347,
>     ...
>   },
>   "repository" : {
>     "id": 1296269,
>     "full_name": "octocat/Hello-World",
>     "owner": {
>       "login": "octocat",
>       "id": 1,
>       ...
>     },
>     ...
>   },
>   "sender": {
>     "login": "octocat",
>     "id": 1,
>     ...
>   }
> }
```

### Sample Payload Endatix
```

Headers
--
content-length | 381
content-type | application/json; charset=utf-8
user-agent | Endatix-HookGun
x-endatix-hook-id | 1300804208097755136
x-endatix-action | created
x-endatix-entity | Submission
x-endatix-event | form_submitted

{
  "Id": 1300804208097755136,
  "EventName": "form_submitted",
  "Action": "created",
  "Payload": {
    "Id": 1300804208097755136,
    "IsComplete": true,
    "JsonData": {
      "name": "Имате ли гъби"
    },
    "Metadata": "string",
    "FormDefinitionId": 1266357680839065600,
    "CurrentPage": 0,
    "CompletedAt": "2024-10-29T12:51:30.953062Z",
    "CreatedAt": "2024-10-29T12:51:30.987541Z"
  }
}
```

## Related Issues
#10 

## Type of Change
Please delete options that are not relevant.

## Checklist

## Screenshots
<img width="1374" alt="image" src="https://github.com/user-attachments/assets/f7e1033d-310f-416c-b8bf-e056d28d8e56">
